BUG: TreeExplainer numerical sensitivity

tylerjereddy opened the pull request:
* Fixes gh-3948

* I drafted the regression test in the matching issue which fails before and passes after the patch discovered by @arhall0 (added as co-author here), that seems to indicate some types were set a bit too small. `pytest --import-mode=append` (full suite) also passes locally on this branch.


## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

tylerjereddy committed:
BUG: TreeExplainer numerical sensitivity

* Fixes gh-3948

* I drafted the regression test in the matching issue which fails before
and passes after the patch discovered by @arhall0 (added as co-author
here), that seems to indicate some types were set a bit too small.
`pytest --import-mode=append` (full suite) also passes locally on this
branch.

Co-authored-by: Aaron R Hall <arhall@lanl.gov>

tylerjereddy commented on tests/explainers/test_tree.py:
```diff
@@ -1898,3 +1898,15 @@ def test_consistency_approximate(expected_result, approximate):
+    rng = np.random.default_rng(0)
+    X = rng.integers(low=0, high=2, size=(n_rows, 90_000)).astype(np.float64)
+    y = rng.integers(low=0, high=2, size=n_rows)
+    clf = sklearn.ensemble.RandomForestClassifier(n_estimators=1, random_state=0)
```
May want to consider parametrizing `n_estimators` over `[1, 100]` -- both caused problems and fail before/pass after--I believe one of the two actually caused a segfault on x86_64 Linux.

There are quite a few issues currently open about additivity check failures, some for 5 years or more, so thorough testing seems worth

CloseChoice replied:
I agree, would be great to have the test parametrized by `n_estimators` as well

CloseChoice reviewed (approved):
Amazing work, thanks for digging so deep and fixing this issue. I leave it to you if you want to parametrize the test.

tylerjereddy committed:
TST: PR 3990 revisions

* `test_gh_3948()` has been additionally parametrized over
`n_estimators`, so that it is a bit more thorough and also
includes a case more similar to the original ticket.

tylerjereddy commented:
I added in the additional test parametrization over `n_estimators`, and confirmed that we see failures with the test on x86_64 Linux before the patch is applied. If you run the test more than once on that platform without the patch, you should start seeing segfaults as well.

CloseChoice commented:
> I added in the additional test parametrization over `n_estimators`, and confirmed that we see failures with the test on x86_64 Linux before the patch is applied. If you run the test more than once on that platform without the patch, you should start seeing segfaults as well.

Thanks for adding.
